### PR TITLE
fix #8802 feat(cirrus): Integrate python nimbus sdk

### LINKED ISSUE #8802: Integrate nimbus sdk python with cirrus
┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-3466)

### PR DESCRIPTION
**Pull request by yashikakhurana**
Because

- We want to use Nimbus SDK (python bindings) to compute enrollment when provided context

This commit

- Add docker instructions to get build for python
- uses Cirrus client to calculate enrollment
- Test cases to verify changes
closes #8802

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -52,3 +52,6 @@ USE_GOOGLE_ANALYTICS=false
 UPLOADS_GS_BUCKET_NAME=
 REMOTE_SETTING_URL=https://firefox.settings.services.mozilla.com/v1/buckets/main/collections/nimbus-web-experiments/records
 REMOTE_SETTING_REFRESH_RATE_IN_SECONDS=10
+APP_ID=fenix
+APP_NAME=fenix
+CHANNEL=developer
```
Comment (yashikakhurana) [inline .env.sample:57]: currently using this to test

File: cirrus/server/cirrus/sdk.py [hunk 2]
```diff
@@ -1,8 +1,33 @@
...
-        return []
+    def __init__(self):
+        self.client = CirrusClient(context)
+
+    def compute_enrollments(self, targeting_context: Dict[str, str]) -> Dict[str, Any]:
+        try:
+            res = self.client.handle_enrollment(  # type: ignore
+                json.dumps(targeting_context)
+            )
+            return json.loads(res)
```
Comment (jaredlockhart) [inline cirrus/server/cirrus/sdk.py:21]: Just a note this `json.loads` could throw as well

File: cirrus/server/cirrus/sdk.py [hunk 3]
```diff
@@ -1,8 +1,33 @@
...
-        self, recipes: List[Dict[str, Any]], targeting_context: Dict[str, str]
-    ) -> List[Dict[str, str]]:
-        return []
+    def __init__(self):
+        self.client = CirrusClient(context)
+
+    def compute_enrollments(self, targeting_context: Dict[str, str]) -> Dict[str, Any]:
+        try:
+            res = self.client.handle_enrollment(  # type: ignore
+                json.dumps(targeting_context)
```
Comment (jaredlockhart) [inline cirrus/server/cirrus/sdk.py:19]: So could this `dumps`

File: cirrus/server/tests/test_sdk.py [hunk 4]
```diff
@@ -1,16 +1,78 @@
...
+        with patch.object(sdk.client, "set_experiments") as mock_set_experiments:
+            sdk.set_experiments(recipes)
+
+        mock_set_experiments.assert_called_once_with(recipes)
+
+    def test_set_experiments_failure(self):
+        sdk = SDK()
+        recipes = '{"data": []}'
+
+        with patch.object(sdk.client, "set_experiments") as mock_set_experiments:
```
Comment (jaredlockhart) [inline cirrus/server/tests/test_sdk.py:62]: Why do we need to mock this out?

**Review by jaredlockhart — APPROVED** [review]
I tested this locally and it built nicely, cached nicely, played around with the SDK in the shell and it was doing stuff!  Lots of questions and ideas here but nothing blocking I'd say let's keep this progress up and continue to iterate.  Great job @yashikakhurana @jeddai 🎉 🎉 🎉 🎉 🎉 🎉

File: cirrus/server/Dockerfile [hunk 5]
```diff
@@ -1,6 +1,43 @@
...
+    python3-venv
+
+RUN git clone https://github.com/mozilla/application-services.git
+RUN if [[ -z "${as_version}" ]] ; then \
+    echo "No application-services version supplied, using lastest commit." \
+    ; else \
+    cd application-services && git fetch \
+    ; fi
+
+WORKDIR /application-services
```
[the remark this answers is not in the document]
Comment (jeddai) [reply cirrus/server/Dockerfile:24]: The clone is cloning into the `/application-services` dir by default, we could move the `WORKDIR` call up above the clone and I think appending `.` to the clone command would have it clone into the current directory. Certainly not a bad plan

[the remark this answers is not in the document]
Comment (jaredlockhart) [reply cirrus/server/Dockerfile:24]: Yeah it's just a bit odd to switch the workdir in the middle of a build

File: cirrus/server/Dockerfile [hunk 6]
```diff
@@ -1,6 +1,43 @@
...
+RUN cargo uniffi-bindgen generate components/nimbus/src/cirrus.udl --language python
+
 # Use official Python runtime as a parent image
 FROM python:3.11.2-slim-buster
 
+# cirrus
+COPY --from=builder /application-services/components/nimbus/src/cirrus.py /application-services/cirrus_sdk.py
+# megazord
+COPY --from=builder /application-services/target/release/libcirrus.so /application-services/libcirrus.so
+ENV PYTHONPATH=$PYTHONPATH:/application-services
```
[the remark this answers is not in the document]
Comment (jeddai) [reply cirrus/server/Dockerfile:39]: Incremental builds are so great

File: cirrus/server/tests/test_sdk.py (hunk 4 above)
Comment (jaredlockhart) [reply cirrus/server/tests/test_sdk.py:62]: Yeah like usually you need to mock out network traffic, but this should be entirely local so we should be able to get away with actually making this call?

File: cirrus/server/cirrus/sdk.py (hunk 3 above)
Comment (yashikakhurana) [reply cirrus/server/cirrus/sdk.py:19]: right, missed that

File: cirrus/server/Dockerfile [hunk 7]
```diff
@@ -1,9 +1,47 @@
...
+    gyp \
+    ninja-build \
+    zlib1g-dev \
+    tclsh \
+    python3-venv
+
+WORKDIR /application-services
+
+RUN git clone https://github.com/mozilla/application-services.git .
+RUN if [[ -z "${as_version}" ]] ; then \
```
Comment (yashikakhurana) [inline cirrus/server/Dockerfile:20]: @jaredlockhart here it using as_version for the future scenario,  cc @jeddai

File: cirrus/server/Dockerfile [hunk 8]
```diff
@@ -1,6 +1,43 @@
+FROM rust:buster AS builder
+
+ARG as_version
```
[the remark this answers is not in the document]
Comment (yashikakhurana) [reply cirrus/server/Dockerfile:3]: https://github.com/mozilla/experimenter/pull/8871/files#r1204869946

File: cirrus/server/Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (yashikakhurana) [reply cirrus/server/Dockerfile:17]: moved up 👍

File: cirrus/server/tests/test_sdk.py (hunk 4 above)
Comment (yashikakhurana) [reply cirrus/server/tests/test_sdk.py]: was writing test cases in the flow, changed :D

**Comment by yashikakhurana** [comment]
merging, for now, if something comes up can file a separate ticket for that :)

**Pull request merged**